### ISSUE TITLE
ipfilter should not trust x-forwarded-for

### ISSUE THREAD

**Issue by captncraig** [pyed/ipfilter#8]
The way your code is handling x-forwarded-for is a bit insecure. Anybody can set the header and be granted access, regardless of their actual location.

I have a new plugin called [realip](https://caddyserver.com/docs/realip) which has just been added to the download page. I would recommend anybody running behind a proxy use that module in addition to ipfilter, so that x-forwarded-for headers are only accepted from a trusted proxy.

**Comment by pyed** [comment]
@fboyer introduced the code for the `x-forwarded-for` header, I'm not sure of the best way to deal with this, looks like if we do not look into `x-forwarded-for` header, using `realip` will be the only way to cover proxied requests, any ideas @fboyer ?

**Comment by fboyer** [comment]
Sorry for the delay in getting back on this topic.

*TL;DR: I'm fine with whatever approach/improvements being proposed here as I am not using caddy server anymore, at least for now, for completely unrelated reasons.*

I believe that the proposal by @captncraig to use *realip* is really scenario dependent, in my case, requests could only come from a trusted proxy sitting in the DMZ of my corporate network that is also used as a TLS termination point. I mean *realip* still puts its trust in a number of proxy servers which could be compromised.

Obviously, if you're server is not really behind a proxy server, meaning that it is accessible from the Interweb, and **if** the filtering rules were discovered by the "attacker" then I understand that it would be fairly easy to circumvent them mostly in the context of one of *ipfilter* primary purpose which is geolocation filtering. In the same way that simply using a VPN service would bypass *ipfilter* even if stricter filtering was in place in the use-case of geolocation filtering. 

This is mostly caused by the following code:
```
func getClientIP(r *http.Request) (net.IP, error) {
	var ip string

	// Use the client ip from the 'X-Forwarded-For' header, if available
	if fwdFor := r.Header.Get("X-Forwarded-For"); fwdFor != "" {
		ip = fwdFor
	} else {
		// Otherwise, get the client ip from the request remote address
		var err error
		ip, _, err = net.SplitHostPort(r.RemoteAddr)
		if err != nil {
			return nil, err
		}
	}

	// Parse the ip address string into a net.IP
	parsedIP := net.ParseIP(ip)
	if parsedIP == nil {
		return nil, errors.New("unable to parse address")
	}

	return parsedIP, nil
}
```

The code was based on the way caddy server deals with the 'X-Forwarded-For' header as seen in the [replacer](https://github.com/mholt/caddy/blob/f4c729bd2205b939f93fad068bbbd7161802f695/middleware/replacer.go#L62-L71) middleware where it's looking for an IP in the 'X-Forwarded-For' header first and foremost and otherwise gets the client IP from the request. (Which could be deemed a bit too naive for the *ipfilter* use case)

Basically what *realip* does is add a validation on the request IP from the request even when 'X-Forwarded-On' is populated. 

In the [discussions](https://github.com/pyed/ipfilter/pull/4) linked to my previous pull request to add support for proxied requests, I initially wanted to add a ```strict``` flag that could have been turned ```on``` to ignore the 'X-Forwarded-For' header during filtering and always use the request IP which would have dealt with the problematic scenario referred to above but we decided not to do it at that point.

Anyway, I am not currently using caddy anymore for completely unrelated reasons but I would encourage you to do a pull request to improve *ipfilter*. 

Anyway, good luck!

PS: I took a quick peek at *realip* and I feel it could easily be rolled into a pull request for *ipfilter* with the addition of one or two configuration options. I am suggesting this due to the overlap around the IP filtering functionality of the two middlewares but that's just my opinion.

**Comment by pyed** [comment]
Ah, you reminded me @fboyer of that `strict` flag you introduced in that PR, and I remember asking you if you think of a scenario where that flag would be useful, I guess `realip` is that scenario, I'll add back that `strict` flag mainly for `realip`.

thank you @fboyer for your enlightening response.

**Comment by captncraig** [comment]
I made realip a seperate plugin because there are other use cases than ip filtering. It is nice to have the real ip in logs, and there are other middlewares that examine the ip address (git does some stuff to authenticate bitbucket among others). 

It felt right to have that logic in a central place so everybody that needs the ip address can just look at what is on the request and not worry about checking headers.

**Comment by fboyer** [comment]
@pyed : I would tend to agree with you unless you wish to completely remove support for proxied request from *ipfilter* and suggest people install the *realip* middleware in front for this specific scenario where requests are proxied. This would simplify your implementation a bit and eliminate this potential insecure configuration but would place a dependency on *realip* to cover all possible scenarios but then again, if you add the ```strict``` flag and set to ```on``` by default, it should at least provide safe/sane defaults.

@captncraig : I believe caddy server offers the ability to log the information in a custom way, see [here](https://caddyserver.com/docs/placeholders), but that might not be what you actually need in terms of logging.

Anyway, depending on what happens and when I decide to go back to caddy server, I might actually leverage both of these modules. :smile: 

Thanks to both of you for your work!

**Comment by pyed** [comment]
@fboyer actually I'm thinking of leaving the defaults to be the same, with `strict` being an optional flag that can be used.

**Comment by pyed** [comment]
we have a new flag now called `strict`, when specified `ipfilter` won't look into `X-Forwarded-For`.

caddyserver.com/docs/ipfilter should be updated anytime soon.
